DylibBinder/DBTypeDeclaration.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using DylibBinder;
+using SwiftReflector;
+using SwiftReflector.Inventory;
+
+namespace DylibBinder {
+	public class DBTypeDeclaration {
+		public DBTypeDeclaration (ProtocolContents protoContents)
+		{
+			Kind = "protocol";
+			Name = protoContents.Name.ToString ();
+			Funcs = new DBFuncs (protoContents);
+			ApplyInouts ();
+		}
+
+		public DBTypeDeclaration (ClassContents classContents)
+		{
+			if (classContents.Name.IsClass)
+				Kind = "class";
+			else if (classContents.Name.IsEnum)
+				Kind = "enum";
+			else
+				Kind = "struct";
+
+			Name = classContents.Name.ToFullyQualifiedName ();
+			Funcs = new DBFuncs (classContents);
+			Properties = new DBProperties (classContents);
+			GenericParameters = new DBGenericParameters (this);
+			InnerTypes = new DBInnerTypes (classContents.Name);
+			AssociatedTypes = new DBAssociatedTypes ();
+			Elements = new DBElements ();
+			ApplyInouts ();
+		}
+
+		public string Kind { get; }
+		public string Name { get; }
+		public string Accessibility { get; } = "Public";
+		public string IsObjC { get; } = "False";
+		public string IsFinal { get; } = "False";
+		public string IsDeprecated { get; } = "False";
+		public string IsUnavailable { get; } = "False";
+
+		public DBFuncs Funcs { get; }
+		public DBProperties Properties { get; }
+		public DBGenericParameters GenericParameters { get; }
+		public DBAssociatedTypes AssociatedTypes { get; }
+		public DBInnerTypes InnerTypes { get; }
+		public DBElements Elements { get; }
+
+		void ApplyInouts ()
+		{
+			foreach (var func in Funcs.Funcs) {
+				ApplyInoutsToParameterLists (func.ParameterLists);
+			}
+			// protocols currently only hold funcs so they should finish here
+			if (Kind == "protocol")
+				return;
+
+			foreach (var property in Properties.Properties) {
+				ApplyInoutsToParameterLists (property.Getter.ParameterLists);
+				if (property.Setter != null) {
+					ApplyInoutsToParameterLists (property.Setter.ParameterLists);
+				}
+			}
+		}
+
+		void ApplyInoutsToParameterLists (DBParameterLists parameterLists)
+		{
+			foreach (var parameterList in parameterLists.ParameterLists) {
+				ApplyInoutsToParameterList (parameterList);
```

Comment:
ApplyInoutsToParameterList coudl this take a params? And we move the loop to that method?

Reply:
Are you suggesting to combine ApplyInoutToParameterLists and ApplyInoutToParameterList to one method?